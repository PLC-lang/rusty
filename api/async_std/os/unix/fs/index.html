<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Unix-specific filesystem extensions."><title>async_std::os::unix::fs - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../../static.files/rustdoc-492a78a4a87dcc01.css"><meta name="rustdoc-vars" data-root-path="../../../../" data-static-root-path="../../../../static.files/" data-current-crate="async_std" data-themes="" data-resource-suffix="" data-rustdoc-version="1.82.0 (f6e511eec 2024-10-15)" data-channel="1.82.0" data-search-js="search-a99f1315e7cc5121.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../../../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../../static.files/main-921df33f47b8780c.js"></script><noscript><link rel="stylesheet" href="../../../../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button><a class="logo-container" href="../../../../async_std/index.html"><img src="https://async.rs/images/logo--hero.svg" alt=""></a></nav><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../../../async_std/index.html"><img src="https://async.rs/images/logo--hero.svg" alt="logo"></a><h2><a href="../../../../async_std/index.html">async_<wbr>std</a><span class="version">1.13.0</span></h2></div><h2 class="location"><a href="#">Module fs</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li></ul></section><h2><a href="../index.html">In async_<wbr>std::<wbr>os::<wbr>unix</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../../index.html">async_std</a>::<wbr><a href="../../index.html">os</a>::<wbr><a href="../index.html">unix</a>::<wbr><a class="mod" href="#">fs</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../../../src/async_std/os/unix/fs.rs.html#1-71">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Unix-specific filesystem extensions.</p>
</div></details><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.DirBuilderExt.html" title="trait async_std::os::unix::fs::DirBuilderExt">DirBuilder<wbr>Ext</a></div><div class="desc docblock-short">Unix-specific extensions to <a href="https://doc.rust-lang.org/1.82.0/std/fs/struct.DirBuilder.html" title="struct std::fs::DirBuilder"><code>fs::DirBuilder</code></a>.</div></li><li><div class="item-name"><a class="trait" href="trait.DirEntryExt.html" title="trait async_std::os::unix::fs::DirEntryExt">DirEntry<wbr>Ext</a></div><div class="desc docblock-short">Unix-specific extension methods for <a href="https://doc.rust-lang.org/1.82.0/std/fs/struct.DirEntry.html" title="struct std::fs::DirEntry"><code>fs::DirEntry</code></a>.</div></li><li><div class="item-name"><a class="trait" href="trait.OpenOptionsExt.html" title="trait async_std::os::unix::fs::OpenOptionsExt">Open<wbr>Options<wbr>Ext</a></div><div class="desc docblock-short">Unix-specific extensions to <a href="https://doc.rust-lang.org/1.82.0/std/fs/struct.OpenOptions.html" title="struct std::fs::OpenOptions"><code>fs::OpenOptions</code></a>.</div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.symlink.html" title="fn async_std::os::unix::fs::symlink">symlink</a></div><div class="desc docblock-short">Creates a new symbolic link on the filesystem.</div></li></ul></section></div></main></body></html>