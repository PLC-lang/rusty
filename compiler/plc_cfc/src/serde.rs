use serde::Deserialize;

// mod json;
/*
interface: local vars (varList)

varList: name, constant, retrain, nonretain, persistent, nonpersistent
varListPlain...

Block {
    position (not needed)
    inputVariables {
        variable {
            connectionPointIn
            formalParameter (requried)
            negated,
            edgeModifierType (do we need this)
            storageModifierType (^)
        }
    }

    InoutVariables {
        connectionPointIn {
            globalId
        }
        connectionPointOut
    }
*/

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
enum PouType {
    Program,
    Function,
    FunctionBlock,
}

#[derive(Debug, Deserialize)]
struct Pou {
    #[serde(rename = "@name")]
    name: String,

    #[serde(rename = "@pouType")]
    pou_type: PouType, // TODO: enum

    #[serde(rename = "@returnType")]
    return_type: Option<String>,

    body: Body,
}

#[derive(Debug, Deserialize)]
struct Body {
    #[serde(rename = "$value")]
    kind: BodyKind,
}

#[derive(Debug, Deserialize)]
enum BodyKind {
    FBD(FunctionBlockDiagram),
}

#[derive(Debug, Deserialize)]
pub struct FunctionBlockDiagram {
    #[serde(rename = "$value")]
    members: Vec<FBDMembers>,
}

// FBD has two sub-objects, commonObjects and fbdObjects; commonObjects consists of mostly UI objects such as vendorElement, do we need these?
#[derive(Debug, Deserialize)]
pub enum FBDMembers {
    #[serde(rename = "block")]
    Block {
        #[serde(rename = "@instanceName")]
        instance_name: Option<String>,

        #[serde(rename = "@typeName")]
        type_name: String,

        #[serde(rename = "@localId")]
        local_id: String,

        #[serde(rename = "@executionOrderId")]
        execution_order_id: Option<String>,

        #[serde(rename = "$value")]
        variables: Vec<VariableKind>,
    },
}

// #[derive(Debug, Deserialize)]
// pub struct Variable {
//     // #[serde(rename = "@formalParameter")]
//     // formal_parameter: String,

//     // #[serde(rename = "@negated")]
//     // negated: bool,

//     // Variable {
//     //     connectionPointIn:
//     // },
//     #[serde(rename = "$value")]
//     kind: VariableKind,
// }

#[derive(Debug, Deserialize)]
#[serde(rename = "camelCase")]
pub enum VariableKind {
    InputVariables(Variable),
    OutVariable(Variable),
}

#[derive(Debug, Deserialize)]

pub struct Variable {
    #[serde(rename = "@formalParameter")]
    formal_parameter: String,

    #[serde(rename = "@negated")]
    negated: bool,

    #[serde(rename = "@connectionPointIn")]
    connection_point_in: Option<ConnectionPoint>,

    #[serde(rename = "@connectionPointOut")]
    connection_point_out: Option<ConnectionPoint>,
}

#[derive(Debug, Deserialize)]

pub struct ConnectionPoint {
    #[serde(rename = "@connection")]
    connection: Connection,
}

#[derive(Debug, Deserialize)]
pub struct Connection {
    #[serde(rename = "@refGlobalId")]
    ref_global_id: Option<String>,

    #[serde(rename = "@globalId")]
    global_id: String,

    #[serde(rename = "@formalParameter")]
    formal_parameter: String,
}

// #[derive(Debug, Deserialize)]
// struct Pou {
//     #[serde(rename = "@name")]
//     name: String,

//     #[serde(rename = "@pouType")]
//     pou_type: PouType, // TODO: enum

//     #[serde(rename = "@returnType")]
//     return_type: Option<String>,

//     body: Body,

//     // TODO: Still missing: external, access, globals
//     // globalVars: Option<Vec<Variable>>,
//     #[serde(rename = "localVariables")]
//     local_vars: Option<Vec<Variable>>,

//     #[serde(rename = "tempVariables")]
//     temp_vars: Option<Vec<Variable>>,

//     #[serde(rename = "inputVariables")]
//     input_vars: Option<Vec<Variable>>,

//     #[serde(rename = "outputVariables")]
//     output_vars: Option<Vec<Variable>>,

//     #[serde(rename = "inOutVariables")]
//     inout_vars: Option<Vec<Variable>>,
// }

// #[derive(Debug, Deserialize)]
// #[serde(rename_all = "camelCase")]
// enum PouType {
//     Program,
//     Function,
//     FunctionBlock,
// }

// #[derive(Debug, Deserialize)]
// struct Variable {
//     #[serde(rename = "@formalParameter")]
//     formal_parameter: String,

//     #[serde(rename = "@negated")]
//     negated: bool,
// }

// #[derive(Debug, Deserialize)]
// pub struct Body {
//     #[serde(rename = "FBD")]
//     fbd: FunctionBlockDiagram,
// }

// #[derive(Debug, Deserialize)]
// pub struct FunctionBlockDiagram {
//     #[serde(rename = "$value")]
//     members: Vec<FBDMembers>,
// }

// // https://stackoverflow.com/a/53366978: Duplicate field error
// // Maybe we have to write our own deserializer (https://serde.rs/impl-deserializer.html)
// #[derive(Debug, Deserialize)]
// enum FBDMembers {
//     #[serde(rename = "block")]
//     Block {
//         #[serde(rename = "@localId")]
//         local_id: String,

//         #[serde(rename = "@executionOrderId")]
//         execution_order_id: String,
//     },

//     #[serde(rename = "inVariable")]
//     InVariable {
//         #[serde(rename = "@localId")]
//         local_id: String,
//     },

//     #[serde(rename = "outVariable")]
//     OutVariable,
// }

// #[derive(Debug, Deserialize)]
// pub struct InVariable {
//     #[serde(rename = "@localId")]
//     local_id: String,
// }

// #[derive(Debug, Deserialize)]
// pub struct OutVariable {}

fn parse() -> Result<(), String> {
    let content = std::fs::read_to_string("res/demo.xml").unwrap();
    let root: Pou = quick_xml::de::from_str(&content).unwrap();
    println!("{root:#?}");

    Ok(())
}

// TODO: Move into folder
mod parse {
    #[test]
    fn xml() {
        let content = crate::parse().unwrap();
        println!("{content:#?}");
    }
}
