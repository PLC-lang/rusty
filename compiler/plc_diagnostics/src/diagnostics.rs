use std::fmt::Display;

use serde::{Deserialize, Serialize};

use plc_ast::ast::AstNode;
use plc_source::{
    source_location::{SourceLocation, SourceLocationFactory},
    SourceCode,
};

pub mod diagnostics_registry;

pub const INTERNAL_LLVM_ERROR: &str = "internal llvm codegen error";

/// a diagnostics severity
#[derive(Default, Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Ignore,
    #[default]
    Info,
    Warning,
    Error,
}

/// The `Diagnostics` struct describes an issue encountered during compile time.
/// The issue is defined by an `error_code` and had a defined `severity`
/// Diagnostic severity can be overridden when being reported.
#[derive(Debug)]
pub struct Diagnostic {
    /// The Description of the error being reported.
    message: String,
    /// Primary location where the diagnostic occurred
    primary_location: SourceLocation,
    /// Seconday locations relevant to the diagnostics
    secondary_locations: Option<Vec<SourceLocation>>,
    /// Error code for reference in the documentation
    error_code: &'static str,
    /// Children of the current diagnostic
    sub_diagnostics: Vec<Diagnostic>,
    /// If the diagnostic is caused by an error, this field contains the original error
    internal_error: Option<anyhow::Error>,
}

impl std::error::Error for Diagnostic {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.internal_error.as_ref().and_then(|it| it.source())
    }
}

impl From<std::io::Error> for Diagnostic {
    fn from(value: std::io::Error) -> Self {
        Diagnostic::new(value.to_string()).with_error_code("E002").with_internal_error(value.into())
    }
}

/// Builder for Diagnostics
impl Diagnostic {
    pub fn new(message: impl Into<String>) -> Self {
        Diagnostic {
            message: message.into(),
            primary_location: SourceLocation::undefined(),
            secondary_locations: Default::default(),
            error_code: "E001", //Default error if none specified
            sub_diagnostics: Default::default(),
            internal_error: Default::default(),
        }
    }

    pub fn with_location<T>(mut self, location: T) -> Self
    where
        T: Into<SourceLocation>,
    {
        self.primary_location = location.into();
        self
    }

    pub fn with_secondary_location<T>(mut self, location: T) -> Self
    where
        T: Into<SourceLocation>,
    {
        self.secondary_locations.get_or_insert_with(Default::default).push(location.into());
        self
    }

    pub fn with_secondary_locations(mut self, locations: Vec<SourceLocation>) -> Self {
        self.secondary_locations.get_or_insert_with(Default::default).extend(locations);
        self
    }

    pub fn with_error_code(mut self, error_code: &'static str) -> Self {
        self.error_code = error_code;
        self
    }

    pub fn with_sub_diagnostic(mut self, diagnostic: Diagnostic) -> Self {
        self.sub_diagnostics.push(diagnostic);
        self
    }

    pub fn with_sub_diagnostics(mut self, diagnostics: Vec<Diagnostic>) -> Self {
        self.sub_diagnostics.extend(diagnostics);
        self
    }

    pub fn with_internal_error(mut self, error: anyhow::Error) -> Self {
        self.internal_error = Some(error);
        self
    }

    pub fn from_serde_error(error: serde_json::Error, source: &SourceCode) -> Self {
        let factory = SourceLocationFactory::for_source(source);
        let line = error.line();
        let column = error.column();

        // remove line, column from message
        let message = error.to_string();
        let message = if let Some(pos) = message.find("at line") {
            message.chars().take(pos).collect()
        } else {
            message
        };
        let range = factory.create_range_to_end_of_line(line, column);
        Diagnostic::new(message).with_error_code("E088").with_location(range)
    }
}

impl PartialEq for Diagnostic {
    fn eq(&self, other: &Self) -> bool {
        self.error_code == other.error_code
            && self.message == other.message
            && self.primary_location == other.primary_location
            && self.secondary_locations == other.secondary_locations
            && self.sub_diagnostics == other.sub_diagnostics
    }
}

impl Eq for Diagnostic {}

impl Display for Diagnostic {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.get_message())?;
        let location = self.get_location();
        if !location.is_undefined() {
            write!(f, " at: {location}")
        } else {
            Ok(())
        }
    }
}

impl Diagnostic {
    pub fn get_message(&self) -> &str {
        self.message.as_str()
    }

    pub fn get_location(&self) -> SourceLocation {
        self.primary_location.clone()
    }

    pub fn get_secondary_locations(&self) -> Option<&[SourceLocation]> {
        self.secondary_locations.as_deref()
    }

    pub fn get_error_code(&self) -> &'static str {
        self.error_code
    }

    pub fn get_sub_diagnostics(&self) -> &[Diagnostic] {
        &self.sub_diagnostics
    }
}

//Helper methods for diagnostics
impl Diagnostic {
    pub fn unexpected_token_found(expected: &str, found: &str, range: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Unexpected token: expected {expected} but found {found}"))
            .with_error_code("E007")
            .with_location(range)
    }

    pub fn missing_function(location: SourceLocation) -> Diagnostic {
        Diagnostic::new("Cannot generate code outside of function context.")
            .with_error_code("E072")
            .with_location(location)
    }

    pub fn codegen_error(message: impl Into<String>, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(message).with_location(location).with_error_code("E071")
    }

    pub fn llvm_error(file: &str, llvm_error: &str) -> Diagnostic {
        Diagnostic::new(format!("{file}: Internal llvm error: {:}", llvm_error)).with_error_code("E005")
    }

    pub fn missing_token(expected_token: &str, range: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Missing expected Token {expected_token}"))
            .with_location(range)
            .with_error_code("E006")
    }

    pub fn invalid_parameter_count(expected: usize, received: usize, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(
            format!(
                "Invalid parameter count. Received {received} parameters while {expected} parameters were expected.",
            )).with_error_code("E032")
            .with_location(location)
    }

    pub fn unknown_type(type_name: &str, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Unknown type: {type_name:}")).with_error_code("E052").with_location(location)
    }

    pub fn unresolved_reference(reference: &str, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Could not resolve reference to {reference:}"))
            .with_error_code("E048")
            .with_location(location)
    }

    pub fn invalid_assignment(right_type: &str, left_type: &str, location: SourceLocation) -> Diagnostic {
        Diagnostic::new(format!("Invalid assignment: cannot assign '{right_type}' to '{left_type}'"))
            .with_error_code("E037")
            .with_location(location)
    }

    pub fn cannot_generate_initializer(variable_name: &str, location: SourceLocation) -> Diagnostic {
        Self::new(format!(
            "Cannot generate literal initializer for '{variable_name}': Value cannot be derived"
        ))
        .with_error_code("E041")
        .with_location(location)
    }

    pub fn cannot_generate_call_statement(operator: &AstNode) -> Diagnostic {
        //TODO: We could probably get a better slice here
        Diagnostic::codegen_error(
            format!("cannot generate call statement for {:?}", operator),
            operator.get_location(),
        )
    }

    pub fn cannot_generate_from_empty_literal(type_name: &str, location: SourceLocation) -> Diagnostic {
        Diagnostic::codegen_error(
            format!("Cannot generate {type_name} from empty literal").as_str(),
            location,
        )
    }
}

// CFC related diagnostics
impl Diagnostic {
    pub fn unnamed_control(location: SourceLocation) -> Diagnostic {
        Diagnostic::new("Unnamed control").with_error_code("E087").with_location(location)
    }
}

#[cfg(test)]
mod tests {
    use codespan_reporting::files::{Location, SimpleFile};

    use crate::{diagnostics::Severity, reporter::clang::ClangFormatDiagnosticReporter};

    #[test]
    fn test_build_diagnostic_msg() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let file = SimpleFile::new("test.st".to_string(), "source".to_string());
        let start = Location { line_number: 4, column_number: 1 };
        let end = Location { line_number: 4, column_number: 4 };
        let res = reporter.build_diagnostic_msg(
            Some(&file),
            Some(&start),
            Some(&end),
            "E001",
            &Severity::Error,
            "This is an error",
        );

        assert_eq!(res, "test.st:4:1:{4:1-4:4}: error[E001]: This is an error");
    }

    #[test]
    fn test_build_diagnostic_msg_equal_start_end() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let file = SimpleFile::new("test.st".to_string(), "source".to_string());
        let start = Location { line_number: 4, column_number: 1 };
        let end = Location { line_number: 4, column_number: 1 };
        let res = reporter.build_diagnostic_msg(
            Some(&file),
            Some(&start),
            Some(&end),
            "E001",
            &Severity::Error,
            "This is an error",
        );

        assert_eq!(res, "test.st:4:1: error[E001]: This is an error");
    }

    #[test]
    fn test_build_diagnostic_msg_no_location() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let file = SimpleFile::new("test.st".to_string(), "source".to_string());
        let res = reporter.build_diagnostic_msg(
            Some(&file),
            None,
            None,
            "E001",
            &Severity::Error,
            "This is an error",
        );

        assert_eq!(res, "test.st: error[E001]: This is an error");
    }

    #[test]
    fn test_build_diagnostic_msg_no_file() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let start = Location { line_number: 4, column_number: 1 };
        let end = Location { line_number: 4, column_number: 4 };
        let res = reporter.build_diagnostic_msg(
            None,
            Some(&start),
            Some(&end),
            "E001",
            &Severity::Error,
            "This is an error",
        );

        assert_eq!(res, "error[E001]: This is an error");
    }

    #[test]
    fn test_build_diagnostic_msg_no_file_no_location() {
        let reporter = ClangFormatDiagnosticReporter::default();
        let res =
            reporter.build_diagnostic_msg(None, None, None, "E001", &Severity::Error, "This is an error");

        assert_eq!(res, "error[E001]: This is an error");
    }
}
