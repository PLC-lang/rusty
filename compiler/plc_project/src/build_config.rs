use diagnostics::Diagnostic;
use plc::Target;
use regex::Captures;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::env;
use std::fs;
use std::path::Path;
use std::path::PathBuf;

use plc::output::FormatOption;

#[derive(Serialize, Deserialize, Debug)]
pub struct LibraryConfig {
    pub name: String,
    pub path: PathBuf,
    pub package: LinkageInfo,
    pub include_path: Vec<PathBuf>,
    #[serde(default = "default_targets")]
    pub architectures: Vec<Target>,
}

/// Targets to use if no other targets have been defined
fn default_targets() -> Vec<Target> {
    vec!["x86_64-linux-gnu".into(), "aarch64-linux-gnu".into()]
}

#[derive(Serialize, Deserialize, Debug, Copy, Clone, PartialEq, Eq)]
pub enum LinkageInfo {
    Copy,
    Local,
    System,
    Static,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ProjectConfig {
    pub name: String,
    pub files: Vec<PathBuf>,
    #[serde(default)]
    pub compile_type: FormatOption,
    #[serde(default)]
    pub output: Option<String>,
    #[serde(default)]
    pub libraries: Vec<LibraryConfig>,
    #[serde(default)]
    pub package_commands: Vec<String>,
}

impl ProjectConfig {
    /// Retuns a project from the given string (in json format)
    /// All environment variables (marked with `$VAR_NAME`) that can be resovled at this time are resolved before the conversion
    pub fn try_parse(content: &str) -> Result<Self, Diagnostic> {
        let content = resolve_environment_variables(content)?;
        serde_json::from_str(&content).map_err(Into::into)
    }

    pub(crate) fn from_file(config: &Path) -> Result<Self, Diagnostic> {
        //read from file
        let content = fs::read_to_string(config)?;

        //convert file to Object
        let project = ProjectConfig::try_parse(&content)?;

        Ok(project)
    }
}

//TODO: I don't think this belongs here
fn resolve_environment_variables(to_replace: &str) -> Result<String, Diagnostic> {
    let pattern = Regex::new(r"\$(\w+)")?;
    let result = pattern.replace_all(to_replace, |it: &Captures| {
        let original = it.get(0).map(|it| it.as_str().to_string()).unwrap();
        if let Some(var) = it.get(1).map(|it| it.as_str()) {
            env::var(var).unwrap_or(original)
        } else {
            original
        }
    });
    Ok(result.replace('\\', r"\\"))
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;
    use std::{env, vec};

    use crate::build_config::default_targets;
    use plc::output::FormatOption;

    use super::LibraryConfig;
    use super::{LinkageInfo, ProjectConfig};

    #[test]
    fn check_build_struct_from_file() {
        let test_project = ProjectConfig {
            name: "MyProject".to_string(),
            files: vec![PathBuf::from("simple_program.st")],
            compile_type: FormatOption::Shared,
            output: Some(String::from("proj.so")),
            libraries: vec![
                LibraryConfig {
                    name: String::from("copy"),
                    path: PathBuf::from("libs/"),
                    package: LinkageInfo::Copy,
                    include_path: vec![PathBuf::from("simple_program.st")],
                    architectures: default_targets(),
                },
                LibraryConfig {
                    name: String::from("nocopy"),
                    path: PathBuf::from("libs/"),
                    package: LinkageInfo::System,
                    include_path: vec![PathBuf::from("simple_program.st")],
                    architectures: default_targets(),
                },
                LibraryConfig {
                    name: String::from("static"),
                    path: PathBuf::from("libs/"),
                    package: LinkageInfo::Static,
                    include_path: vec![PathBuf::from("simple_program.st")],
                    architectures: default_targets(),
                },
                LibraryConfig {
                    name: String::from("withTargets"),
                    path: PathBuf::from("libs/"),
                    package: LinkageInfo::Static,
                    include_path: vec![PathBuf::from("simple_program.st")],
                    architectures: vec!["myArch".into(), "myArch2".into()],
                },
            ],
            package_commands: vec![],
        };
        let proj = ProjectConfig::try_parse(
            r#"
            {
                "name": "MyProject",
                "files" : [
                    "simple_program.st"
                ],
                "compile_type" : "Shared",
                "output": "proj.so",
                "libraries" : [
                    {
                        "name" : "copy",
                        "path" : "libs/",
                        "package" : "Copy",
                        "include_path" : [
                            "simple_program.st"
                        ]
                    },
                    {
                        "name" : "nocopy",
                        "path" : "libs/",
                        "package" : "System",
                        "include_path" : [
                            "simple_program.st"
                        ]
                    },
                    {
                        "name" : "static",
                        "path" : "libs/",
                        "package" : "Static",
                        "include_path" : [
                            "simple_program.st"
                        ]
                    },
                    {
                        "name" : "withTargets",
                        "path" : "libs/",
                        "package" : "Static",
                        "include_path" : [
                            "simple_program.st"
                        ],
                        "architectures": ["myArch", "myArch2"]
                    }
                ]
            }
        "#,
        )
        .unwrap();

        assert_eq!(test_project.name, proj.name);
        assert_eq!(test_project.files, proj.files);
        assert_eq!(test_project.compile_type, proj.compile_type);
        assert_eq!(test_project.output, proj.output);
        let proj_lib = proj.libraries;
        let testproj_lib = test_project.libraries;
        assert_eq!(testproj_lib[0].name, proj_lib[0].name);
        assert_eq!(testproj_lib[0].path, proj_lib[0].path);
        assert_eq!(testproj_lib[0].package, proj_lib[0].package);
        assert_eq!(testproj_lib[0].include_path, proj_lib[0].include_path);
        assert_eq!(testproj_lib[1].name, proj_lib[1].name);
        assert_eq!(testproj_lib[1].path, proj_lib[1].path);
        assert_eq!(testproj_lib[1].package, proj_lib[1].package);
        assert_eq!(testproj_lib[1].include_path, proj_lib[1].include_path);
    }

    #[test]
    fn project_creation_resolves_environment_vars() {
        //Add env
        env::set_var("test_var", "test_value");
        let proj = ProjectConfig::try_parse(
            r#"
            {
                "name" : "$test_var",
                "files" : [
                    "simple_program.st"
                ]
            }
        "#,
        )
        .unwrap();

        assert_eq!("test_value", &proj.name);
    }
}
