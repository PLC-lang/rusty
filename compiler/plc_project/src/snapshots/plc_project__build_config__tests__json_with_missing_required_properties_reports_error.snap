---
source: compiler/plc_project/src/build_config.rs
expression: diag.to_string()
---
plc_json__invalid: missing field `name`  at: :6:9:{6:9-6:100}: 
